samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "ble_connectivity_manager.h"
+
+#include <bluetooth/gatt_dm.h>
+#include <bluetooth/scan.h>
+#include <bluetooth/services/lbs.h>
+
+#include <platform/CHIPDeviceLayer.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+using namespace chip;
+
+BLEConnectivityManager BLEConnectivityManager::sBLEConnectivityManager;
+
+BT_SCAN_CB_INIT(scan_cb, BLEConnectivityManager::FilterMatch, NULL, NULL, NULL);
+
+BT_CONN_CB_DEFINE(conn_callbacks) = { .connected = BLEConnectivityManager::ConnectionHandler,
+				      .disconnected = BLEConnectivityManager::DisconnectionHandler };
+
+static const struct bt_gatt_dm_cb discovery_cb = { .completed = BLEConnectivityManager::DiscoveryCompletedHandler,
+						   .service_not_found = BLEConnectivityManager::DiscoveryNotFound,
+						   .error_found = BLEConnectivityManager::DiscoveryError };
+
+static struct bt_conn_le_create_param *create_param = BT_CONN_LE_CREATE_CONN;
+
+void BLEConnectivityManager::FilterMatch(struct bt_scan_device_info *device_info,
+					 struct bt_scan_filter_match *filter_match, bool connectable)
+{
+	/* Limit the number of devices that can be scanned. */
+	if (GetBLEConnectivityManager().mScannedDevicesCounter >= kMaxScannedDevices) {
+		return;
+	}
+
+	/* Verify the device address to make sure that the scan result will be handled only once for every device. */
+	for (int i = 0; i < GetBLEConnectivityManager().mScannedDevicesCounter; i++) {
+		if (memcmp(device_info->recv_info->addr, &GetBLEConnectivityManager().mScannedDevices[i].mAddr,
+			   sizeof(GetBLEConnectivityManager().mScannedDevices[i].mAddr)) == 0) {
+			return;
+		}
+	}
+
+	memcpy(&GetBLEConnectivityManager().mScannedDevices[GetBLEConnectivityManager().mScannedDevicesCounter].mAddr,
+	       device_info->recv_info->addr,
+	       sizeof(GetBLEConnectivityManager()
+			      .mScannedDevices[GetBLEConnectivityManager().mScannedDevicesCounter]
+			      .mAddr));
+	memcpy(&GetBLEConnectivityManager()
+			.mScannedDevices[GetBLEConnectivityManager().mScannedDevicesCounter]
+			.mConnParam,
+	       device_info->conn_param,
+	       sizeof(GetBLEConnectivityManager()
+			      .mScannedDevices[GetBLEConnectivityManager().mScannedDevicesCounter]
+			      .mConnParam));
+
+	/* TODO: obtain more information from advertising data to help user distinguish the devices (e.g. BT name). */
+
+	GetBLEConnectivityManager().mScannedDevicesCounter++;
+}
+
+void BLEConnectivityManager::ConnectionHandler(struct bt_conn *conn, uint8_t conn_err)
+{
+	const bt_addr_le_t *addr = bt_conn_get_dst(conn);
```

Comment:
Here should be added a validation to check if we should use this part of the code because it also will be used during the commissioning process.

Reply:
As we discussed f2f, there is not much we can do to avoid it, as while being in the central role we cannot clearly say when we started or not (like it is done for peripheral role in case of advertising). For now checking if the address is present on a list of connected devices and returning in case it is not there seems to be good enough.